core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1103 @@
+	ops, prelude::*,
+	convert::{TryInto, TryFrom},
+	cmp::{min, max, Ordering}
+};
+use codec::{Encode, Decode};
+use crate::traits::{
+	SaturatedConversion, CheckedSub, CheckedAdd, Bounded, UniqueSaturatedInto, Saturating, Zero,
+};
+
+macro_rules! implement_per_thing {
+	($name:ident, $test_mod:ident, [$($test_units:tt),+], $max:tt, $type:ty, $upper_type:ty, $title:expr $(,)?) => {
+		/// A fixed point representation of a number between in the range [0, 1].
+		///
+		#[doc = $title]
+		#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug, Ord, PartialOrd))]
+		#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq)]
+		pub struct $name($type);
+
+		impl $name {
+			/// Nothing.
+			pub fn zero() -> Self { Self(0) }
+
+			/// `true` if this is nothing.
+			pub fn is_zero(&self) -> bool { self.0 == 0 }
+
+			/// Everything.
+			pub fn one() -> Self { Self($max) }
+
+			/// Consume self and deconstruct into a raw numeric type.
+			pub fn deconstruct(self) -> $type { self.0 }
+
+			/// Return the scale at which this per-thing is working.
+			pub const fn accuracy() -> $type { $max }
+
+			/// From an explicitly defined number of parts per maximum of the type.
+			///
+			/// This can be called at compile time.
+			pub const fn from_parts(parts: $type) -> Self {
+				Self([parts, $max][(parts > $max) as usize])
+			}
+
+			/// Converts from a percent. Equal to `x / 100`.
+			///
+			/// This can be created at compile time.
+			pub const fn from_percent(x: $type) -> Self {
+				Self([x, 100][(x > 100) as usize] * ($max / 100))
+			}
+
+			/// Return the product of multiplication of this value by itself.
+			pub fn square(self) -> Self {
+				// both can be safely casted and multiplied.
+				let p: $upper_type = self.0 as $upper_type * self.0 as $upper_type;
+				let q: $upper_type = $max as $upper_type * $max as $upper_type;
+				Self::from_rational_approximation(p, q)
+			}
+
+			/// Converts a fraction into `Permill`.
+			#[cfg(feature = "std")]
+			pub fn from_fraction(x: f64) -> Self { Self((x * ($max as f64)) as $type) }
+
+			/// Approximate the fraction `p/q` into a per-thing fraction. This will never overflow.
+			///
+			/// The computation of this approximation is performed in the generic type `N`. Given
+			/// `M` as the data type that can hold the maximum value of this per-thing (e.g. u32 for
+			/// perbill), this can only work if `N == M` or `N: From<M> + TryInto<M>`.
+			pub fn from_rational_approximation<N>(p: N, q: N) -> Self
+				where N: Clone + Ord + From<$type> + TryInto<$type> + ops::Div<N, Output=N>
+			{
+				// q cannot be zero.
+				let q = q.max((1 as $type).into());
+				// p should not be bigger than q.
+				let p = p.min(q.clone());
+
+				let factor = (q.clone() / $max.into()).max((1 as $type).into());
+
+				// q cannot overflow: (q / (q/$max)) < 2 * $max. p < q hence p also cannot overflow.
+				// this implies that $type must be able to fit 2 * $max.
+				let q_reduce: $type = (q / factor.clone())
```

Comment:
if max is u8::max_value(), type is u8 and upper type is u16 then factor is 39 and q/factor == 256 and doesn't fit into u8.

Maybe when playing with u8::max_value and edge stuff approximation can make inequalities wrong :-/

Reply:
yes indeed better to saturate but I think maybe we can also just not permit it. Tbh I don't see any value in the current `Percent` which I implemented just for demonstration. For the first round, we can enforce `$type: From<u32>` (based on wasm being 32bit it also doesn't really bring us much in terms of memory or performance if we don arithmetic in 16 bit) and later refactor and harden if we actually need sth like `Percent`.

Reply:
I just checked and:

- Things like percent can be actually created. Just had to change some tests. 
- The example you gave will not be accepted since we already discussed that for any per-thing, `$type` must be able to fit `2 * $max`. What you described is a `per_u8` with `$type = u8` which does not fit into this description.